File: banchus-app/src/main/java/pe/nanamochi/banchus/controller/web/ScreenshotController.java
package pe.nanamochi.banchus.controller.web;

import java.io.IOException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import pe.nanamochi.banchus.service.StorageService;
import pe.nanamochi.banchus.service.UserService;

@Slf4j
@RestController
@RequestMapping("/web")
@RequiredArgsConstructor
public class ScreenshotController {
  private final UserService userService;
  private final StorageService storageService;

  @PostMapping("/osu-screenshot.php")
  public ResponseEntity<String> uploadScreenshot(
      @RequestParam(value = "u", required = false) String username,
      @RequestParam(value = "p", required = false) String passwordMd5,
      @RequestParam(value = "v", required = false) Integer endpointVersion,
      @RequestParam(value = "ss", required = false) MultipartFile screenshotFile) {
    return userService
        .login(username, passwordMd5)
        .map(
            user -> {
              if (screenshotFile == null || screenshotFile.isEmpty()) {
                return ResponseEntity.badRequest().body("Empty file");
              }

              if (screenshotFile.getSize() > 4 * 1024 * 1024) {
                return ResponseEntity.status(HttpStatus.CONTENT_TOO_LARGE).body("File too large");
              }

              if (endpointVersion != null && endpointVersion != 1) {
                log.warn(
                    "Incorrect endpoint version for {}: v{}", user.getUsername(), endpointVersion);
              }

              try {
                String filename = storageService.saveScreenshot(screenshotFile.getBytes());
                log.debug("Screenshot saved: {} by user: {}", filename, user.getUsername());
                return ResponseEntity.ok(filename);
              } catch (IOException e) {
                log.error("Physical error saving screenshot from {}", user.getUsername(), e);
                return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).<String>build();
              }
            })
        .orElseGet(() -> ResponseEntity.status(HttpStatus.UNAUTHORIZED).build());
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/controller/client/web/ScreenshotController.kt
package pe.nanamochi.banchus.controller.client.web

import com.github.michaelbull.result.*
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import pe.nanamochi.banchus.database.entity.User
import pe.nanamochi.banchus.infrastructure.security.AuthenticatedUser
import pe.nanamochi.banchus.service.StorageService

@RestController
@RequestMapping("/web")
class ScreenshotController(private val storageService: StorageService) {
    private val log = LoggerFactory.getLogger(javaClass)

    @PostMapping("/osu-screenshot.php")
    fun uploadScreenshot(
        @AuthenticatedUser user: User,
        @RequestParam(value = "v", required = false) endpointVersion: Int?,
        @RequestParam(value = "ss", required = false) screenshotFile: MultipartFile?,
    ): ResponseEntity<String> {
        endpointVersion
            ?.takeIf { it != 1 }
            ?.let { log.warn("Incorrect endpoint version for {}: v{}", user.username, it) }

        val result = binding {
            val file =
                screenshotFile?.takeIf { !it.isEmpty }.toResultOr { HttpStatus.BAD_REQUEST }.bind()

            file
                .takeIf { it.size <= 4 * 1024 * 1024 }
                .toResultOr {
                    log.warn(
                        "User {} upload rejected: file too large ({} bytes)",
                        user.username,
                        file.size,
                    )
                    HttpStatus.CONTENT_TOO_LARGE
                }
                .bind()

            val filename =
                runCatching { file.bytes }
                    .mapError {
                        log.error(
                            "Failed to read bytes from multipart file for user {}",
                            user.username,
                        )
                        HttpStatus.INTERNAL_SERVER_ERROR
                    }
                    .andThen { bytes ->
                        storageService.saveScreenshot(bytes).mapError {
                            HttpStatus.INTERNAL_SERVER_ERROR
                        }
                    }
                    .bind()

            filename
        }

        return result.mapBoth(success = { ResponseEntity.ok(it) }, failure = { ResponseEntity(it) })
    }
}
